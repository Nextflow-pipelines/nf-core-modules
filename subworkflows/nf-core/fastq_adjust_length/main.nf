// TODO nf-core: If in doubt look at other nf-core/subworkflows to see how we are doing things! :)
//               https://github.com/nf-core/modules/tree/master/subworkflows
//               You can also ask for help via your pull request or on the #subworkflows channel on the nf-core Slack workspace:
//               https://nf-co.re/join
// TODO nf-core: A subworkflow SHOULD import at least two modules

include { SEQKIT_SUBSEQ                         } from '../../../modules/nf-core/seqkit/subseq/main.nf'
include { SEQKIT_STATS as SEQKIT_STATS_RAW      } from '../../../modules/nf-core/seqkit/stats/main.nf'
include { SEQKIT_STATS as SEQKIT_STATS_ADJUSTED } from '../../../modules/nf-core/seqkit/stats/main.nf'

workflow FASTQ_ADJUST_LENGTH {

    take:
    // TODO nf-core: edit input (take) channels
    ch_reads // channel: [ val(meta), [ fastq ] ]
    ch_aparams

    main:

    ch_versions = Channel.empty()

    // TODO nf-core: substitute modules here for the modules of your subworkflow

    //ch_reads.

    SEQKIT_STATS_RAW ( ch_reads )
    ch_versions = ch_versions.mix(SEQKIT_STATS_RAW.out.versions.first())

    SEQKIT_STATS_RAW.out.stats
    .map { meta, stats ->
        Channel.fromPath(stats)
        .splitCsv(header: true, sep: "\t")
        .map { row ->
            meta = meta + ['num_seqs_raw': row.num_seqs, 'num_seqs_raw': row.min_len, 'min_len_raw': row.min_len, 'max_len_raw': row.max_len ]
            return [ meta.id, [meta, stats]]
        }
    }
    .set { ch_stats_raw }

    ch_reads
    .map { meta, reads ->
        return [meta.id, [meta, reads]]
    }
    .join(ch_stats_raw, by:0)
    .map { id, ch_items ->
        (reads, stats) = ch_items
        meta = stats[0]
        return [meta, reads[1]]
    }
    .combine(ch_aparams)
    .map { meta, reads, aparams ->
        meta.read1_astart  = aparams[0]
        meta.read1_alength = aparams[1]
        meta.read2_astart  = aparams[2]
        meta.read2_alength = aparams[3]
        return [meta, reads]
    }
    .set { input_fastq }

    SEQKIT_SUBSEQ ( input_fastq )
    ch_versions = ch_versions.mix(SEQKIT_SUBSEQ.out.versions.first())

    SEQKIT_SUBSEQ.out.adjusted_resds
    .set { ch_reads_adjuted }

    SEQKIT_STATS_ADJUSTED ( ch_reads_adjuted )
    ch_versions = ch_versions.mix(SEQKIT_STATS_ADJUSTED.out.versions.first())

    SEQKIT_STATS_ADJUSTED.out.stats
    .map { meta, stats ->
        Channel.fromPath(stats)
        .splitCsv(header: true, sep: "\t")
        .map { row ->
            meta = meta + ['num_seqs_adjusted': row.num_seqs, 'num_seqs_adjusted': row.min_len, 'min_len_adjusted': row.min_len, 'max_len_adjusted': row.max_len ]
            return [ meta.id, [meta, stats]]
        }
    }
    .set { ch_stats_adjusted }

    ch_reads_adjuted
    .map { meta, reads ->
        return [meta.id, [meta, reads]]
    }
    .join(ch_stats_adjusted, by:0)
    .map { id, ch_items ->
        (reads, stats) = ch_items
        meta = stats[0]
        return[meta, reads[1]]
    }
    .set { output_fastq }

    emit:
    // TODO nf-core: edit emitted channels
    fastq    = output_fastq                    // channel: [ val(meta), [ reads ] ]

    versions = ch_versions                     // channel: [ versions.yml ]
}
