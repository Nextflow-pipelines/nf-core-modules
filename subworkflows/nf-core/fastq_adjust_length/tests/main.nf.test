// todo nf-core: once you have added the required tests, please run the following command to build this file:
// nf-core subworkflows test fastq_adjust_length
nextflow_workflow {

    name "test subworkflow fastq_adjust_length"
    script "../main.nf"
    workflow "fastq_adjust_length"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_adjust_length"
    // todo nf-core: add tags for all modules used within this subworkflow. example:
    tag "seqkit"
    tag "seqkit/subseq"
    tag "seqkit/stats"


    // todo nf-core: change the test name preferably indicating the test-data and file-format used
    test("sarscov2 - fastq - single_end") {

        when {
            workflow {
                """
                // todo nf-core: define inputs of the workflow here. example:
                input[0] = channel.of([
                        [ id:'test', single_end:true ], // meta map
                        [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkifexists: true) ]
                        ])
                input[1] = channel.of([ [1, 75, null, null] ])
                """
            }
        }

        then {
            assertall(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
                //todo nf-core: add all required assertions to verify the test output.
            )
        }
    }

    test("sarscov2 - fastq - paired_end") {
        when {
            workflow {
                """
                // TODO nf-core: define inputs of the workflow here. Example:
                input[0] = Channel.of([
                        [ id:'test', single_end:false ], // meta map
                        [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                          file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true)]
                        ])
                input[1] = Channel.of([ [1, 75, 1, 75] ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
                //TODO nf-core: Add all required assertions to verify the test output.
            )
        }
    }
}
