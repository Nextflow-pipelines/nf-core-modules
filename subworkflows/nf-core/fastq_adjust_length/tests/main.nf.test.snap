{
    "sarscov2 - fastq - paired_end": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "fastq": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-14T22:32:15.9263"
    }
}